test/cpp_extensions/cpp_c10d_extension.cpp

```diff
@@ -0,0 +1,122 @@
+#include "cpp_c10d_extension.hpp"
+
+#include <map>
+
+namespace c10d {
+
+ProcessGroupTest::WorkTest::~WorkTest() {}
+
+bool ProcessGroupTest::WorkTest::isCompleted() {
+  return true;
+}
+
+bool ProcessGroupTest::WorkTest::isSuccess() const {
+  return true;
+}
+
+bool ProcessGroupTest::WorkTest::wait() {
+  return true;
+}
+
+ProcessGroupTest::ProcessGroupTest(int rank, int size)
+    : ProcessGroup(rank, size) {}
+
+ProcessGroupTest::~ProcessGroupTest() {}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupTest::broadcast(
+    std::vector<at::Tensor>& tensors,
+    const BroadcastOptions& opts) {
+  return std::make_shared<ProcessGroupTest::WorkTest>();
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupTest::allreduce(
+    std::vector<at::Tensor>& tensors,
+    const AllreduceOptions& opts) {
+  return std::make_shared<ProcessGroupTest::WorkTest>();
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupTest::allreduce_coalesced(
+      std::vector<at::Tensor>& tensors,
+      const AllreduceCoalescedOptions& opts) {
+  throw std::runtime_error("ProcessGroupTest does not support allreduce_coalesced");
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupTest::reduce(
+    std::vector<at::Tensor>& tensors,
+    const ReduceOptions& opts) {
+  throw std::runtime_error("ProcessGroupTest does not support reduce");
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupTest::allgather(
+    std::vector<std::vector<at::Tensor>>& outputTensors,
+    std::vector<at::Tensor>& inputTensors,
+    const AllgatherOptions& opts) {
+  throw std::runtime_error("ProcessGroupTest does not support allgather");
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupTest::allgather_coalesced(
+    std::vector<std::vector<at::Tensor>>& outputTensorLists,
+    std::vector<at::Tensor>& inputTensors,
+    const AllgatherOptions& opts) {
+  throw std::runtime_error("ProcessGroupTest does not support allgather_coalesced");
+}
```

Comment:
We (+@mrshenli) have recently made changes to ProcessGroup API and decided that we'll only require implementing allgather_base which gathers just 1 tensor (use one tensor as input and one tensor as output that would contain all gathered tensors concatenated), e.g. see https://github.com/pytorch/pytorch/pull/31892.
However, this process is not completed yet.

Could you implement allgather_base here, otherwise your code would break the build?
Also, we already allow clients not to implement allgather_coalesced, so you can drop this implementation.

Just FYI for now: the same is true for allreduce_coalesced, but we haven't started the process of deprecating it yet.

Reply:
ok. I will update the code accordingly.